autoscheduler/frontend/src/redux/reducers/courseCards.ts

```diff
@@ -43,34 +43,97 @@ const initialCourseCardArray: CourseCardArray = {
+): CourseCardArray {
+  // Determine new number of cards created
+  const numCardsCreated = Math.max(state.numCardsCreated, index + 1);
+
+  const newState: CourseCardArray = { numCardsCreated };
+
+  for (let i = 0; i < numCardsCreated; i++) {
+    if (state[i] || i === index) {
+      const shouldExpand = i === index;
+      newState[i] = {
+        ...state[i],
+        ...(shouldExpand ? courseCard : {}),
+        collapsed: !shouldExpand,
+      };
+    }
+  }
+  return newState;
+}
+
 // reducer
 export default function courseCards(
   state: CourseCardArray = initialCourseCardArray, action: CourseCardAction,
 ): CourseCardArray {
   switch (action.type) {
     case ADD_COURSE_CARD: {
       const newCardIdx = action.idx ?? state.numCardsCreated;
+      // If new card is explicitly expanded, perform necessary state changes
+      if (action.courseCard.collapsed === false) {
+        return getStateAfterExpanding(state, newCardIdx, action.courseCard);
+      }
+      // New card isn't supposed to be expanded, simply add it
       return {
         ...state,
         [newCardIdx]: action.courseCard,
         numCardsCreated: Math.max(state.numCardsCreated, newCardIdx + 1),
       };
     }
-    case REMOVE_COURSE_CARD:
-      return {
-        ...state,
-        [action.index]: undefined,
-      };
-    case UPDATE_COURSE_CARD:
+    case REMOVE_COURSE_CARD: {
+      // Delete desired card
+      let newState: CourseCardArray = state;
+
+      // If the deleted card was expanded, expand the one below it
+      const wasExpanded = state[action.index].collapsed === false;
+      if (wasExpanded) {
+        let newExpandedIdx: number;
+        for (let i = action.index - 1; i >= 0; i--) {
+          if (state[i]) {
+            newExpandedIdx = i;
+            break;
+          }
+        }
+        // If no cards were below the deleted card, expand the card above it
+        if (newExpandedIdx === undefined) {
+          for (let i = action.index + 1; i <= state.numCardsCreated; i++) {
+            if (state[i]) {
+              newExpandedIdx = i;
+              break;
+            }
+          }
+        }
+        if (newExpandedIdx !== undefined) {
+          newState = getStateAfterExpanding(state, newExpandedIdx, undefined);
+        }
+      }
+      newState = { ...newState, [action.index]: undefined };
+      return newState;
+    }
+    case UPDATE_COURSE_CARD: {
+      // if card doesn't exist, don't update
       if (!state[action.index]) return state;
+      // if card was expanded, collapse other cards
+      if (action.courseCard.collapsed === false && state[action.index]?.collapsed !== false) {
+        return getStateAfterExpanding(state, action.index, action.courseCard);
+      }
+
+      // otherwise just update this card
       return {
         ...state,
         [action.index]: { ...state[action.index], ...action.courseCard },
         numCardsCreated: Math.max(state.numCardsCreated, action.index + 1),
       };
+    }
```

Comment:
Don't think you need the braces around this case block, as you don't declare anything